doc/decisions/header_include.md

```diff
@@ -0,0 +1,131 @@
+>
+> 3. A preprocessing directive of the form
+>
+>    ```c
+>    #include "q-char-sequence" new-line
+>    ```
+>
+>    causes the replacement of that directive by the entire contents of the source file identified by the specified sequence between the `"` delimiters. The named source file is searched for in an implementation-defined manner. If this search is not supported, or if the search fails, the directive is reprocessed as if it read
+>
+>    ```c
+>    #include <h-char-sequence> new-line
+>    ```
+>
+>    with the identical contained sequence (including `>` characters, if any) from the original directive.
+>
+> 4. A preprocessing directive of the form
+>
+>    ```c
+>    #include pp-tokens new-line
+>    ```
+>
+>    (that does not match one of the two previous forms) is permitted. The preprocessing tokens after include in the directive are processed just as in normal text. (Each identifier currently defined as a macro name is replaced by its replacement list of preprocessing tokens.) The directive resulting after all replacements shall match one of the two previous forms. The method by which a sequence of preprocessing tokens between a `<` and a `>` preprocessing token pair or a pair of `"` characters is combined into a single header name preprocessing token is implementation-defined.
+>
+> ### Definitions:
+>
+> - h-char: any member of the source character set except the new-line character and `>`
+> - q-char: any member of the source character set except the new-line character and `"`
+
+In short, the whole search process of the `#include` mechansim is implementation-defined.
+However, because `#include` works (mostly) like a literal copy-paste, deciding between `""` and `<>` within a public header file can affect user code.
+
+## Assumptions
+
+- Modern Compilers work like [GCC](https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html):
+  - `#include "[header]"` treats `[header]` as file path relative to the current file (as defined by the standard, if such a file doesn't exist there is a fallback to the `<>` behavior)
+  - `#include <[header]>` treats `[header]` as file path relative to one of the pre-defined include-paths
+
+## Considered Alternatives
+
+- Let developers decide:
+
+  This would just create inconsistent code.
+
+- Always use `<>`:
+
+  This can make things more difficult than necessary. We wouldn't utilize the common "relative path" behaviour. Therefore, user code must have all the same directories in the include-paths as our code (translated to their installed counterparts).
+
+- Put all headers into one directory and use the same relative layout as will be installed. Then use `""`:
+
+  This makes for an inconvenient development experience. It is also pretty hard to achieve for plugins and private headers would clutter the include-directory.
+
+## Decision
+
+The rules for including headers are:
+
+- Including another header from the current directory is done with:
```

Comment:
I wouldn't introduce such a rule. It is very hard to enforce it. And what is the benefit?

Reply:
> It is very hard to enforce it.

Code review is always an option, but it also shouldn't be too hard to write a script that checks this, or add a custom query to our lgtm.com build job.

> And what is the benefit?

1. It clearly shows whether or not a header is part of the same library or not.
2. You need to type less than. The alternative would be always having `<elektra/mylib/header.h>`.
3. Generally I'd say in C code the convention is `<>` is for external/system headers and `""` is for local headers. So when I see a `<>` I expect it to come from somewhere else.
4. When you use `""` your editor/IDE doesn't need to know about the correct include path. Since the header will be in the same directory the header can be found without extra setup. With `<>` an editor/IDE that isn't setup explicitly, will just look in the system folders. So it might find a version of the header, but it could be the wrong one (installed vs. local source).

Ideally I'd like to use `""` for including all Elektra headers from within the code base. Then there can't be any confusion. But this would require careful coordination between the source tree and the install locations and also the use of `../`, so I made this compromise.